File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();

        for(int i=0; i<asteroids.length; i++){
            int asteroid = asteroids[i];
            boolean destroyed = false;

            while(!stack.isEmpty() && stack.peek() > 0 && asteroid < 0){
                if(stack.peek() > Math.abs(asteroid)){
                    destroyed = true;
                    break;
                }
                else if(stack.peek() == Math.abs(asteroid)){
                    stack.pop();
                    destroyed = true;
                    break;
                }else
                {
                    stack.pop();
                }
            }

            if(!destroyed) stack.push(asteroid);
        }

        int[] result = new int[stack.size()];
        for(int i=result.length-1; i>=0; i--){
            result[i] = stack.pop();
        }

        return result;
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> stack;

        for(int i=0; i<asteroids.size(); i++){
            int asteroid = asteroids[i];
            bool destroyed = false;

            while(!stack.empty() && stack.top() > 0 && asteroid < 0){
                if(stack.top() > abs(asteroid)){
                    destroyed = true;
                    break;
                }else if(stack.top() == abs(asteroid)){
                    stack.pop();
                    destroyed = true;
                    break;
                }
                else{
                    stack.pop();
                }
            }

            if(!destroyed) stack.push(asteroid);
        }

        vector<int> result(stack.size());
        for(int i=result.size()-1; i>=0; i--){
            int num = stack.top();
            result[i] = num;
            stack.pop();
        }

        return result;
    }
};